include/boost/decimal/detail/cmath/log10.hpp

```diff
@@ -26,21 +27,119 @@ template <typename T>
 constexpr auto log10_impl(T x) noexcept
     BOOST_DECIMAL_REQUIRES(detail::is_decimal_floating_point_v, T)
 {
-    // TODO(ckormanyos) Actually this is eintirely preliminary. The implementation
-    // of log10 will/should-be entirely re-worked with specific base-10-relevant details.
-
-    // TODO(ckormanyos) Handle non-normal varguments.
-
-    // TODO(ckormanyos) Put in a basic check for pure powers of 10, resulting
-    // in an exact result.
-
-    // Starting point: See implementation of log() and adapt to the following series.
-    // Series[Log[10] Log[10, (1 + (z/2))/(1 - (z/2))], {z, 0, 17}]
-
-    return log(x) / numbers::ln10_v<T>;
+    constexpr T zero { 0, 0 };
+    constexpr T one  { 1, 0 };
+
+    T result { };
+
+    const auto fpc = fpclassify(x);
+
+    if (fpc == FP_ZERO)
+    {
+        result = -std::numeric_limits<T>::infinity();
+    }
+    else if (signbit(x) || (fpc == FP_NAN))
+    {
+        result = std::numeric_limits<T>::quiet_NaN();
+    }
+    else if (fpc == FP_INFINITE)
+    {
+        result = std::numeric_limits<T>::infinity();
+    }
+    else
+    {
+        int exp10val { };
+
+        using representation_type = std::conditional_t<std::is_same<T, decimal32>::value, std::uint32_t,
+                                    std::conditional_t<std::is_same<T, decimal64>::value, std::uint64_t, detail::uint128>>;
+
+        const representation_type gn { frexp10(x, &exp10val) };
+
+        const remove_trailing_zeros_return<representation_type>
+            zeros_removal
+            {
+                remove_trailing_zeros(gn)
+            };
```

Comment:
Couldn't all of this use of `representation_type` be replaced by `auto`? Then we wouldn't need to fuss with around with adding the typedef.

Reply:
> Couldn't all of this use of `representation_type` be replaced by `auto`?

Yes. Let me run that through lickety split.